apps/prairielearn/elements/pl-excalidraw/pl-excalidraw.py

```diff
@@ -0,0 +1,218 @@
+import base64
+import json
+from pathlib import Path
+from typing import List
+
+import chevron
+import lxml.html
+import prairielearn as pl
+from lxml.html import HtmlElement
+from typing_extensions import assert_never
+
+ATTR_ANSWER_NAME = "answers-name"
+ATTR_WIDTH = "width"
+ATTR_HEIGHT = "height"
+ATTR_SOURCE_FILE_NAME = "source-file-name"
+ATTR_SOURCE_DIRECTORY = "directory"
+ATTR_GRADABLE = "gradable"
+
+
+SOURCE_DIRECTORY_MAP = {
+    "serverFilesCourse": "server_files_course_path",
+    "clientFilesCourse": "client_files_course_path",
+    "clientFilesQuestion": "client_files_question_path",
+    ".": "question_path",
+}
+
+# -----------------------------------------------------------------------------
+# The logic below is derived from the combinations at
+# https://github.com/PrairieLearn/PrairieLearn/pull/9900#discussion_r1685163461
+
+
+def is_answer_name_required(gradable: bool) -> bool:
+    # All cases marked (ANS)
+    return gradable
+
+
+def is_widget_editable(panel: str, gradable: bool, editable: bool) -> bool:
+    # All cases where `(RW)` is marked. data["editable"] must also be True
+    return panel == "question" and gradable and editable
+
+
+def is_source_file_name_required(panel: str, gradable: bool, fresh: bool) -> bool:
+    # All cases where `NoSource: Raise an error`
+    match panel:
```

Comment:
Can you write this as a boolean expression that gets returned instead of using the cases? That style is a little bit cleaner.